source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,500 @@
+    dispatch_loop->is_destroying = true;
+
+    /* make sure the loop is running so we can schedule a last task. */
+    s_run(event_loop);
+
+    /* cancel outstanding tasks */
+    dispatch_async_and_wait(dispatch_loop->dispatch_queue, ^{
+      aws_mutex_lock(&dispatch_loop->synced_data.lock);
+      dispatch_loop->synced_data.current_thread_id = aws_thread_current_thread_id();
+      dispatch_loop->synced_data.is_executing = true;
+      aws_mutex_unlock(&dispatch_loop->synced_data.lock);
+
+      aws_task_scheduler_clean_up(&dispatch_loop->scheduler);
+
+      while (!aws_linked_list_empty(&dispatch_loop->synced_data.cross_thread_tasks)) {
+          struct aws_linked_list_node *node = aws_linked_list_pop_front(&dispatch_loop->synced_data.cross_thread_tasks);
+          struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+          task->fn(task, task->arg, AWS_TASK_STATUS_CANCELED);
+      }
+
+      while (!aws_linked_list_empty(&dispatch_loop->local_cross_thread_tasks)) {
+          struct aws_linked_list_node *node = aws_linked_list_pop_front(&dispatch_loop->local_cross_thread_tasks);
+          struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+          task->fn(task, task->arg, AWS_TASK_STATUS_CANCELED);
+      }
+
+      aws_mutex_lock(&dispatch_loop->synced_data.lock);
+      /* The entries in the scheduled_services are already put on the apple dispatch queue. It would be a bad memory
+       * access if we destroy the entries here. We instead setting a cancel flag to cancel the task when the
+       * dispatch_queue execute the entry. */
+      struct aws_linked_list_node *iter = NULL;
+      for (iter = aws_linked_list_begin(&dispatch_loop->synced_data.scheduling_state.scheduled_services);
+           iter != aws_linked_list_end(&dispatch_loop->synced_data.scheduling_state.scheduled_services);
+           iter = aws_linked_list_next(iter)) {
+          struct scheduled_service_entry *entry = AWS_CONTAINER_OF(iter, struct scheduled_service_entry, node);
+          entry->cancel = true;
+      }
+      dispatch_loop->synced_data.suspended = true;
+      dispatch_loop->synced_data.is_executing = false;
+      aws_mutex_unlock(&dispatch_loop->synced_data.lock);
+    });
+
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Releasing Dispatch Queue.", (void *)event_loop);
+    aws_ref_count_release(&dispatch_loop->ref_count);
+}
+
+static int s_wait_for_stop_completion(struct aws_event_loop *event_loop) {
+    (void)event_loop;
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_run(struct aws_event_loop *event_loop) {
+    struct dispatch_loop *dispatch_loop = event_loop->impl_data;
+
+    aws_mutex_lock(&dispatch_loop->synced_data.lock);
+    if (dispatch_loop->synced_data.suspended) {
+        AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Starting event-loop thread.", (void *)event_loop);
+        dispatch_resume(dispatch_loop->dispatch_queue);
+        dispatch_loop->synced_data.suspended = false;
```

Comment:
Shouldn't this schedule a service on the off-chance that there were scheduled tasks at the time of suspension?

Reply:
Updated to schedule a service on event loop resume.